lib/parser/canonicalize-omp.cc

```diff
@@ -0,0 +1,127 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
```

Comment:
lib/parser doesn't seem to me to be the right directory for semantic rewriting passes.

Reply:
This is really a syntactic operation, so I think this is an appropriate place for it. If it weren't for the probably of labeled DO statements, the parser would be taking care of associating DO loops with OMP directives.

For the same reason, CanonicalizeDo should be in `lib/parser`. It's already in `namespace Fortran::parser`, which is confusing for something in `lib/semantics`. Both of these operations are about constructing a good parse tree and don't depend on any semantic analysis.

Reply:
I would be okay with having *both* of those rewrites in lib/parser, but not just one of them.  And they should be driven by the parser, not `Semantics::Perform()`, if they're not in lib/semantics.

Reply:
Yes, that makes sense. I suggest that for this PR, leave `canonicalize-omp.cc` in `lib/parser`. When it is merged I'll move `canonicalize-do.cc` in there too and add a new member function in `parser::Parsing` that calls both canonicalize functions. I'll call that from the driver and remove the calls from `Semantics::Perform()`.

Reply:
Excellent.  Thanks.

Reply:
> It may work to commute to work by driving your car in reverse, but that might be good engineering.

`canonicalize-do` was written with the assertion that label resolution must have happened. If labels are missing, misplaced, duplicated, etc., etc., the canonicalizer may choke hard, corrupting the parse tree and/or providing little or no feedback to the user.

This is not the only case where there is a phase ordering in semantics checking and creating/modifying front-end data structures. In general, rewriting the syntax tree blindly and before any sort of correctness testing is performed seems like the wrong approach.

Reply:
Sounds like the dependence on label resolution will force DO loop canonicalization to remain in semantics.  If this code also depends on successful label resolution checks, it will also need to go into semantics.  Otherwise, if we're absolutely sure that it's safe to run without semantics, it can stay in the parser.

Reply:
Actually, this code depends on the successful result (`structured do-loops`) of the DO loop canonicalization.

Reply:
Based on the above discussions and after talking with Steve, I've moved this code to `lib/semantics` for now.